src/elexmodel/handlers/data/Featurizer.py

```diff
@@ -0,0 +1,123 @@
+        # effect columns for aggregation.
+        original_fixed_effect_columns = df[self.fixed_effects]
+        return pd.concat(
+            [
+                pd.get_dummies(
+                    df,
+                    columns=self.fixed_effects,
+                    prefix=self.fixed_effects,
+                    prefix_sep="_",
+                    dtype=np.int64,
+                    drop_first=drop_first,
+                ),
+                original_fixed_effect_columns,
+            ],
+            axis=1,
+        )
+
+    def featurize_fitting_data(self, fitting_data, center_features=True, add_intercept=True):
+        """
+        Featurize the data that the model is fitted on.
+        """
+        # make copy of fitting_data, since we do not want to change the original data
+        new_fitting_data = fitting_data.copy()
+        self.center_features = center_features
+        self.add_intercept = add_intercept
+
+        if self.center_features:
+            self._center_features(new_fitting_data)
+
+        if self.add_intercept:
+            self._add_intercept(new_fitting_data)
+
+        if len(self.fixed_effects) > 0:
+            # drop_first is True for fitting_data (reporting_units) since we want to avoid the design matrix with
+            # expanded fixed effects to be linearly dependent
+            new_fitting_data = self._expand_fixed_effects(new_fitting_data, drop_first=True)
+            # we save the expanded fixed effects to be able to add fixed effects that are not in the heldout_data (nonreporting_units)
+            # as a zero column and to be able to specify the order of the expanded fixed effect when fitting the model
+            self.expanded_fixed_effects = [
+                x
+                for x in new_fitting_data.columns
+                if x.startswith(tuple([fixed_effect + "_" for fixed_effect in self.fixed_effects]))
+            ]
+
+        # all features that the model will be fit on
+        self.complete_features = ["intercept"] + self.features + self.expanded_fixed_effects
+
+        return new_fitting_data[self.complete_features]
+
+    def featurize_heldout_data(self, heldout_data):
+        """
+        Featurize the data that the model will be applied on.
+        """
+        new_heldout_data = heldout_data.copy()
+
+        if self.center_features:
+            self._center_features(new_heldout_data)
+
+        if self.add_intercept:
+            self._add_intercept(new_heldout_data)
+
+        if len(self.fixed_effects) > 0:
+            missing_expanded_fixed_effects = []
+            new_heldout_data = self._expand_fixed_effects(new_heldout_data, drop_first=False)
+            # if all units from one fixed effect are reporting they will not appear in the heldout_data (nonreporting_units) and won't
+            # get a column when we expand the fixed effects on that dataframe. Therefore we add those columns with zero
```

Comment:
similar here - can you clarfy that non-reporting units is one example of heldout_data? I think the only other is the conformalization data in the CI training/fitting?

Reply:
Same as above, let me know

Reply:
same

Reply:
done!